Approved/Initializers.md

```diff
@@ -0,0 +1,175 @@
+<!-- 
+title: Initializers for qubit allocations
+description: Allows to initialize types containing qubits and qubits to a certain state
+author: Bettina Heim
+date: 10/03/2020
+-->
+
+# Proposal
+
+The proposal is to introduce a more general notion of initializers for qubit allocations. It achieves to support three important functionalities as part of using- and borrowing statements:
+- Creating instances of user defined types containing qubits
+- Initializing the allocated qubits with a certain operation upon allocation
+- Automatic allocation and de-allocation of temporary qubits used during initialization
+
+# Justification
+
+The current qubit allocation is limited. While all of the three bullets above can be achieved with the current support, expressing them is somewhat verbose. This proposal introduces a much more convenient and readable syntax and automation around these scenarios. 
+
+The proposed modification allows for nesting of initializers; i.e. the initialized values can be used as part of other initializers in the same allocation, making it possible to automatically manage temporary qubit values that are only used as part of the initialization. 
+
+Furthermore, this proposal makes it possible to support defining custom constructors for user defined types that contain qubits, without requiring that the necessary qubits are passed as arguments to the constructor.    
+
+# Description
+
+This proposal formalizes/introduces a couple of concepts that are outlined in the following. See [this section](#overview-over-different-kinds-of-initializer-expressions) for an overview based on examples of the different kinds of initializers and initializer expressions.
+
+### *Allocatable types*:   
+A type is allocatable if it either is one of the built-in allocatable types, or if it contains one or more items which are allocatable. There are two built-in allocatable types: the type Qubit and the type Qubit[], or more generally n-dimensional rectangular Qubit arrays (as suggested [here](https://github.com/microsoft/qsharp-language/issues/39)). 
+
+### *Initializers for allocatable types*:    
+Initializers are special callables that return values of an allocatable type. Initializers are *neither* operations nor functions. They are their own distinct type with no subtyping relation to operation or function types; they cannot be used when a value of operation or function type is required. Initializers can only be used as part of [qubit allocation statements](https://github.com/microsoft/qsharp-language/blob/main/Specifications/Language/2_Statements/QuantumMemoryManagement.md#quantum-memory-management).   
```

Comment:
Having three different types of callable in Q# is a lot, especially since initializer names overlap with both type names and type constructor names. The automatic conversion from e.g. `LittleEndian(Qubit[])` to `LittleEndian(Int)` is also a little confusing - when I see `LittleEndian(3)`, I have to remember that:

1. It's in an initializer expression, so `LittleEndian` refers to the initializer, not the constructor, which takes a different argument type.
2. The `Int` maps back to the length of a `Qubit` array, even though `Qubit` doesn't appear anywhere in the expression.

I think there is a lot of cognitive overhead needed to interpret this syntax.